### PR TITLE
[Backport] 8262355: Support for AVX-512 opmask register allocation.

### PR DESCRIPTION
**Pull request by JinZhonghui**
Summary: 8262355: Support for AVX-512 opmask register allocation.

Test Plan: ci jtreg

Reviewed-by: JoshuaZhuwj

Issue: https://github.com/dragonwell-project/dragonwell11/issues/607

### CODE REVIEW

**Comment by JinZhonghui** [comment]
@JoshuaZhuwj  Please help to review this.

**Review by JoshuaZhuwj — COMMENTED** [review]
Zhonghui, Thanks for your contributing.

As our offline communication,  VectorMaskGen, LoadVectorMasked and StoreVectorMasked should be introduced since they are the basis of mask support.

See my other comments inline.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 1]
```diff
@@ -9438,6 +9454,13 @@ void Assembler::evpblendmq (XMMRegister dst, KRegister mask, XMMRegister nds, XM
   emit_int16(0x64, (0xC0 | encode));
 }
 
+void Assembler::bzhiq(Register dst, Register src1, Register src2) {
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/assembler_x86.cpp:9473]: Removed and do it in 8261553.

File: src/hotspot/cpu/x86/assembler_x86.hpp [hunk 2]
```diff
@@ -2110,6 +2112,8 @@ class Assembler : public AbstractAssembler  {
   void shlxl(Register dst, Register src1, Register src2);
   void shlxq(Register dst, Register src1, Register src2);
 
+  void bzhiq(Register dst, Register src1, Register src2);
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/assembler_x86.hpp:2116]: Removed and do it in 8261553.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 3]
```diff
@@ -11185,66 +11410,181 @@ instruct rep_stos_large(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
...
+       $$emit$$"jge     L_sloop\n\t"
+       $$emit$$"# L_end:\n\t"
+    } else {
+       $$emit$$"xorq    rax, rax\t# ClearArray:\n\t"
+       $$emit$$"rep     stosq\t# Store rax to *rdi++ while rcx--"
+    }
+  %}
+  ins_encode %{
+    __ clear_mem($base$$Register, $cnt$$Register, $zero$$Register,
+                 $tmp$$XMMRegister, true);
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_64.ad]: Done.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 4]
```diff
@@ -11135,10 +11297,73 @@ instruct rep_stos(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
...
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+// Large ClearArray non-AVX512.
+instruct rep_stos_large(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
                         Universe dummy, rFlagsReg cr)
 %{
-  predicate(((ClearArrayNode*)n)->is_large());
+  predicate(UseAVX <=2 && ((ClearArrayNode*)n)->is_large());
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_64.ad:11366]: Done.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 5]
```diff
@@ -11135,10 +11297,73 @@ instruct rep_stos(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
   ins_pipe(pipe_slow);
 %}
 
-instruct rep_stos_large(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero, 
+// Small ClearArray AVX512 non-constant length.
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_64.ad:11300]: Done.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 6]
```diff
@@ -11135,10 +11297,73 @@ instruct rep_stos(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
...
+       $$emit$$"jge     L_sloop\n\t"
+       $$emit$$"# L_end:\n\t"
+    } else {
+       $$emit$$"rep     stosq\t# Store rax to *rdi++ while rcx--\n\t"
+    }
+    $$emit$$"# DONE"
+  %}
+  ins_encode %{
+    __ clear_mem($base$$Register, $cnt$$Register, $zero$$Register,
+                 $tmp$$XMMRegister, false);
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_64.ad]: Done.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 7]
```diff
@@ -11073,13 +11235,13 @@ instruct MoveL2D_reg_reg(regD dst, rRegL src) %{
   ins_pipe( pipe_slow );
 %}
 
-
-// =======================================================================
-// fast clearing of an array
+// Fast clearing of an array
+// Small ClearArray non-AVX512.
 instruct rep_stos(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_64.ad:11240]: Done.

File: src/hotspot/cpu/x86/x86_32.ad [hunk 8]
```diff
@@ -11685,13 +11789,76 @@ instruct rep_stos(eCXRegI cnt, eDIRegP base, regD tmp, eAXRegI zero, Universe du
   %}
   ins_encode %{
     __ clear_mem($base$$Register, $cnt$$Register, $zero$$Register,
-                 $tmp$$XMMRegister, false);
+                 $tmp$$XMMRegister, false, knoreg);
   %}
   ins_pipe( pipe_slow );
 %}
 
+// Small ClearArray AVX512 non-constant length.
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/x86_32.ad:11797]: Done.

File: src/hotspot/cpu/x86/gc/z/zBarrierSetAssembler_x86.cpp [hunk 9]
```diff
@@ -519,13 +537,14 @@ class ZSaveLiveRegisters {
     _xmm_registers.sort(xmm_compare_register_size);
 
     // Stack pointer must be 16 bytes aligned for the call
-    _spill_offset = _spill_size = align_up(xmm_spill_size + gp_spill_size, 16);
+    _spill_offset = _spill_size = align_up(xmm_spill_size + gp_spill_size + opmask_spill_size + frame::arg_reg_save_area_bytes, 16);
```
[the remark this answers is not in the document]
Comment (JinZhonghui) [reply src/hotspot/cpu/x86/gc/z/zBarrierSetAssembler_x86.cpp]: frame::arg_reg_save_area_bytes is introduced from '8233299: Implementation: JEP 365: ZGC on Windows'.
Removed.

**Comment by JoshuaZhuwj** [comment]
Just for future reference.
See the new version of "AVX-512 opmask register allocation" at https://github.com/dragonwell-project/dragonwell11/pull/703

**Pull request closed without merging**